src/facade/reply_builder.h

```diff
@@ -167,54 +167,119 @@ class SinkReplyBuilder {
   bool send_active_ : 1;
 };
 
-// TMP: New version of reply builder that batches not only to a buffer, but also iovecs.
+// Base class for all reply builders. Offer a simple high level interface for controlling output
+// modes and sending basic response types.
 class SinkReplyBuilder2 {
+  struct GuardBase {
+    bool prev;
+    SinkReplyBuilder2* rb;
+  };
+
+ public:
+  constexpr static size_t kMaxInlineSize = 32;
+  constexpr static size_t kMaxBufferSize = 8192;
+
   explicit SinkReplyBuilder2(io::Sink* sink) : sink_(sink) {
   }
 
   // Use with care: All send calls within a scope must keep their data alive!
   // This allows to fully eliminate copies for batches of data by using vectorized io.
-  struct ReplyScope {
-    explicit ReplyScope(SinkReplyBuilder2* rb) : prev_scoped(rb->scoped_), rb(rb) {
-      rb->scoped_ = true;
+  struct ReplyScope : GuardBase {
+    explicit ReplyScope(SinkReplyBuilder2* rb) : GuardBase{std::exchange(rb->scoped_, true), rb} {
     }
-    ~ReplyScope() {
-      if (!prev_scoped) {
-        rb->scoped_ = false;
-        rb->FinishScope();
-      }
+
+    ~ReplyScope();
+  };
+
+  // Reduce number of send calls by aggregating responses in a buffer. Prefer ReplyScope
+  // if it's conditions are met.
```

Comment:
its

Reply:
how these are different? can you point out explicitly?

Reply:
Batching can be enabled if we send pipeline responses or computed data. In that case we want to minimize the number of send calls irrespective to the number of copies. 

Scoping assumes all the string_views keep valid. This means we can hold onto the data for as long as possible (by not copying, but referencing as an iovec). We decide what to do with it only once the scope ends (or the piece buffer fills up)

Reply:
```c++
ReplyAggregator guard();
StartArray(n);
for (cmd: commands)
  SendBulkString(RunCommand(cmd));
```

will batch responses to the buffer (because we don't keep RunCommand output alive) and result likely in a single vector to call `Send({&buffer})`

however

```c++
string strs[] = {};

ReplyScope scope(this); // we assure we can safely reference strs as long as guard is alive
StartArray(n);
for (str: strs)
  SendBulkString(str)
```

will batch vecors and the result is something like this

`Send({&buffer[] /* array len,  str len, crlf */ , &str1, &bufer[]  /* str len, crlf */  , &str2, ....})`

meaning neither of the strings got copied

Reply:
I think this is a perfect explanation that is worth adding to before of these classes because the semantics are subtle.

Reply:
I'm not suggesting to remove array functions and use scopes everywhere, most of them will be used internally. 

For example, this is SendScoredArray. The scope means that SendBulkString can just safely keep a vector to the string until the function is over (or even longer if there is a higher-level scope)

```c++
void RedisReplyBuilder::SendScoredArray(absl::Span<const std::pair<std::string, double>> arr,
                                        bool with_scores) {
  ReplyScope scope(this);
  StartArray((with_scores && !IsResp3()) ? arr.size() * 2 : arr.size());
  for (const auto& [str, score] : arr) {
    if (with_scores && IsResp3())
      StartArray(2);
    SendBulkString(str);
    if (with_scores)
      SendDouble(score);
  }
}
```